lib/spack/spack/spec.py

```diff
@@ -773,18 +796,37 @@ def constrain(self, other):
             elif k not in self:
                 self[k] = other[k]
                 changed = True
+            # Check that propagation
+            if self[k] == other[k]:
+                for i in range(len(other[k])):
+                    if self[k][i].propagate != other[k][i].propagate:
+                        raise UnsatisfiableCompilerFlagSpecError(
+                            self[k][i].propagate, other[k][i].propagate
+                        )
         return changed
 
     @staticmethod
     def valid_compiler_flags():
         return _valid_compiler_flags
 
     def copy(self):
-        clone = FlagMap(None)
-        for name, value in self.items():
-            clone[name] = value
+        clone = FlagMap(self.spec)
+        for name, compiler_flag in self.items():
+            clone[name] = compiler_flag
         return clone
 
+    def add_flag(self, flag_type, value, propagation):
```

Comment:
Perhaps add a docstring here to explain the method and arguments since this is a brand new method (e.g., see the docstring for `satisfies()` at ~line 399.